### PR TITLE
Allow user to set DNS lookup family as part of bootstrap config

### PR DESCRIPTION
**Pull request by abhide**
Updates: #3323

### CODE REVIEW

File: internal/envoy/bootstrap.go [hunk 1]
```diff
@@ -84,6 +84,9 @@ type BootstrapConfig struct {
 	// referenced in the configuration actually exist. This option is for
 	// testing only.
 	SkipFilePathCheck bool
+
+	// DNSLookupFamily
```
Comment (abhide) [inline internal/envoy/bootstrap.go:88]: oops missed this out. will fix in next patch

File: cmd/contour/bootstrap.go [hunk 2]
```diff
@@ -35,5 +35,6 @@ func registerBootstrap(app *kingpin.Application) (*kingpin.CmdClause, *envoy.Boo
 	bootstrap.Flag("envoy-key-file", "Client key filename for Envoy secure xDS gRPC communication.").Envar("ENVOY_KEY_FILE").StringVar(&config.GrpcClientKey)
 	bootstrap.Flag("namespace", "The namespace the Envoy container will run in.").Envar("CONTOUR_NAMESPACE").Default("projectcontour").StringVar(&config.Namespace)
 	bootstrap.Flag("xds-resource-version", "The versions of the xDS resources to request from Contour.").Default("v3").StringVar((*string)(&config.XDSResourceVersion))
+	bootstrap.Flag("dns-lookup-family", "DNS Resolution Policy. Either v4, v6 or auto").StringVar(&config.DNSLookupFamily)
```
Comment (abhide) [inline cmd/contour/bootstrap.go:38]: naming is hard! :)

Comment (sunjayBhatia) [inline cmd/contour/bootstrap.go:38]: This addition also needs some docs here: https://projectcontour.io/docs/v1.12.0/configuration/#flags (source file https://github.com/projectcontour/contour/blob/d31d2b6fe22097e44ab55924e4707927c5084807/site/docs/main/configuration.md#flags)

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]
I think to fully fix https://github.com/projectcontour/contour/issues/3323 we need to allow configurability of the DNS lookup family on `ExtensionService` clusters as well. Right now I *believe* we don't set anything on those clusters so they would do ipv6/ipv4 lookups as well. We might want to make that configurable per `ExtensionService` or take this global default, not sure! Could be done in another PR though

As final validation for this we should deploy a Contour + Envoys and configure some `Ingresses`, and `HTTPProxies` with `ExtensionServices` and monitor the k8s DNS logs.

File: cmd/contour/bootstrap.go (hunk 2 above)
Comment (abhide) [reply cmd/contour/bootstrap.go]: Thanks @sunjayBhatia . Fixed this in next patchset.

**Comment by abhide** [comment]
> I think to fully fix #3323 we need to allow configurability of the DNS lookup family on `ExtensionService` clusters as well. Right now I _believe_ we don't set anything on those clusters so they would do ipv6/ipv4 lookups as well. We might want to make that configurable per `ExtensionService` or take this global default, not sure! Could be done in another PR though
> 
> As final validation for this we should deploy a Contour + Envoys and configure some `Ingresses`, and `HTTPProxies` with `ExtensionServices` and monitor the k8s DNS logs.

@sunjayBhatia Thanks for giving further direction here. Currently, seems like we set `dns-lookup-family` for Kubernetes `externalName` service via Contour. Yes, we need to extend it for clusters. Will look into that next. Would be good if we can merge this PR if things look good and address the remaining items in a next PR. Will keep #3323 open till its correctly fixed.

File: cmd/contour/bootstrap.go [hunk 3]
```diff
@@ -35,5 +35,6 @@ func registerBootstrap(app *kingpin.Application) (*kingpin.CmdClause, *envoy.Boo
 	bootstrap.Flag("envoy-key-file", "Client key filename for Envoy secure xDS gRPC communication.").Envar("ENVOY_KEY_FILE").StringVar(&config.GrpcClientKey)
 	bootstrap.Flag("namespace", "The namespace the Envoy container will run in.").Envar("CONTOUR_NAMESPACE").Default("projectcontour").StringVar(&config.Namespace)
 	bootstrap.Flag("xds-resource-version", "The versions of the xDS resources to request from Contour.").Default("v3").StringVar((*string)(&config.XDSResourceVersion))
+	bootstrap.Flag("dns-lookup-family", "DNS Resolution Policy for Envoy -> Contour cluster lookup. Either v4, v6 or auto").StringVar(&config.DNSLookupFamily)
```
Comment (youngnick) [inline cmd/contour/bootstrap.go:38]:
```suggestion
	bootstrap.Flag("dns-lookup-family", "Defines what DNS Resolution Policy to use for Envoy -> Contour cluster name lookup. Either v4, v6 or auto").StringVar(&config.DNSLookupFamily)
```

**Review by youngnick — COMMENTED** [review]
One small wording suggestion, but other than that, LGTM.

File: cmd/contour/bootstrap.go (hunk 3 above)
Comment (abhide) [reply cmd/contour/bootstrap.go]: Fixed

File: internal/envoy/bootstrap.go (hunk 1 above)
Comment (sunjayBhatia) [inline internal/envoy/bootstrap.go:88]: Should we add a descriptive comment here? At least to make the linter etc. happy (though it is a bit redundant and an internal struct field)

**Review by sunjayBhatia — APPROVED** [review]

**Review by skriss — APPROVED** [review]
LGTM

File: internal/envoy/bootstrap.go (hunk 1 above)
Comment (youngnick) [reply internal/envoy/bootstrap.go:88]: Yes, we absolutely should have a more descriptive comment here. Better to have too much documentation than not enough.

Comment (abhide) [reply internal/envoy/bootstrap.go]: done

**Pull request merged**